File: pi/lib/herdr_workflow/prompts.py
"""Pure(ish) prompt/argument building. Only local filesystem reads for extension/plugin discovery — no subprocess."""
import json
import shlex
from pathlib import Path

from . import paths

UNRESTRICTED_ROLES = {"planner", "worker"}
ONE_SHOT_ROLES = {"recovery", "archive"}
# Herdr extensions loaded by explicit --extension flag, skip in discovery to avoid double-loading
HERDR_EXTENSIONS = {"herdr-telemetry", "herdr-workflow"}
PI_EXTENSION_DIRS = [paths.AGENT_DIR / "extensions", paths.AGENT_DEF_DIR / "extensions", Path.home() / ".config" / "pi" / "extensions"]

ROLE_TOOLS = {
    "planner": "read,bash,edit,write",
    "triage": "read,bash,edit,write",
    "recovery": "read,bash,edit,write",
    "worker": "read,bash,edit,write",
    "security-verifier": "read,bash",
    "agents-verifier": "read,bash",
    "quality-verifier": "read,bash",
    "performance-verifier": "read,bash",
    "openspec-verifier": "read,bash",
    "test-verifier": "read,bash",
    "archive": "read,bash",
}


def is_one_shot(role):
    return role in ONE_SHOT_ROLES


def role_env(role, change):
    return ["--env", f"HERDR_ROLE={role}", "--env", f"HERDR_CHANGE_ID={change}"]


def discover_extensions():
    """Discover all extension files from standard pi locations."""
    extensions = {}
    for directory in PI_EXTENSION_DIRS:
        if not directory.is_dir():
            continue
        for entry in sorted(directory.iterdir()):
            if entry.suffix in {".ts", ".js", ".mjs"}:
                name = entry.stem
                if name not in extensions:
                    extensions[name] = str(entry)
    return extensions


def load_plugin_assignments():
    """Load plugin-assignments.json if it exists."""
    path = paths.AGENT_DIR / "plugin-assignments.json"
    if path.exists():
        try:
            return json.loads(path.read_text())
        except (json.JSONDecodeError, OSError):
            print(f"warning: corrupt {path}, starting fresh")
    return {"plugins": []}


def save_plugin_assignments(assignments):
    """Save plugin-assignments.json."""
    path = paths.AGENT_DIR / "plugin-assignments.json"
    path.write_text(json.dumps(assignments, indent=2) + "\n")


def resolve_exclusions(config, role):
    """Resolve excluded extension names for a given role."""
    excluded = set()
    plugins = config.get("plugins", {})
    if "exclude_extensions" in plugins:
        excluded.update(plugins["exclude_extensions"])
    role_config = plugins.get("roles", {}).get(role, {})
    if "exclude_extensions" in role_config:
        excluded.update(role_config["exclude_extensions"])
    return excluded


def role_prompt(role, change, verification_round=None, workflow_type=None, task=None):
    request = f".herdr-workflow/{change}/request.md"
    restricted = " Complete assigned role in this Pi process then stop — do not stay active waiting for next step. You will be notified when needed again. Do not invoke another agent executable or invoke another agent or pane. Use herdr-workflow for required workflow handoff exactly as specified."
    persistent = " Complete this round in this Pi process, then go idle and wait in this same process for the next round's prompt — do not exit, restart, or start unrelated work. Do not invoke another agent executable or invoke another agent or pane. Use herdr-workflow for required workflow handoff exactly as specified."
    if role == "planner":
        return f"Planner for OpenSpec change {change}. Read {request}, explore repository context, and discuss unclear requirements with developer. When asked to propose, write proposal, design, tasks, and delta spec scenarios under openspec/changes/{change}/. Submit with herdr-workflow phase --repo . --change {change} proposed; fix PLAN_REJECTED feedback before finishing."
    if role == "worker":
        if workflow_type == "no-openspec":
            description = f" Implement this change: {task}" if task else ""
            return f"Worker for {change}. Use chat for scope, progress, and blockers. No task checklist to read — signal completion by running herdr-workflow verify --repo . --change {change} once the change is applied.{description}"
        return f"Worker for {change}. Follow loaded skill and use chat for scope, progress, and blockers. Apply approved plan. Mark each OpenSpec task [x] only after focused validation; verification rejects unfinished tasks."
    if role == "triage":
        reviews = f".herdr-workflow/{change}/reviews"
        return f"Silent triage for round {verification_round}. Read {reviews}/round-{verification_round}-triage-input.json only. Select minimum needed reviewers and assign relevant files or hunks. Write {reviews}/round-{verification_round}-triage.json, then run herdr-workflow dispatch-verifiers --repo . --change {change}. No chat output." + persistent
    if role == "security-verifier":
        context = f".herdr-workflow/{change}/reviews/round-{verification_round}-security-verifier-context.md"
        report = f".herdr-workflow/{change}/reviews/round-{verification_round}-security-verifier.findings.jsonl"
        return f"Silent security verifier for {change} round {verification_round}. Read {context}. Review changed trust boundaries for introduced injection, auth, secret, crypto, and input-validation defects. Write JSONL findings plus final PASS/FAIL verdict to {report}, then run herdr-workflow verification-result --repo . --change {change} --role security-verifier. No chat output." + persistent
    if role == "agents-verifier":
        context = f".herdr-workflow/{change}/reviews/round-{verification_round}-agents-verifier-context.md"
        report = f".herdr-workflow/{change}/reviews/round-{verification_round}-agents-verifier.findings.jsonl"
        return f"Silent AGENTS instructions verifier for {change} round {verification_round}. Read {context}. Check changed code only against applicable AGENTS.md and CLAUDE.md instructions. Write JSONL findings plus final PASS/FAIL verdict to {report}, then run herdr-workflow verification-result --repo . --change {change} --role agents-verifier. No chat output." + persistent
    if role == "quality-verifier":
        context = f".herdr-workflow/{change}/reviews/round-{verification_round}-quality-verifier-context.md"
        report = f".herdr-workflow/{change}/reviews/round-{verification_round}-quality-verifier.findings.jsonl"
        return f"Silent code quality verifier for {change} round {verification_round}. Read {context}. Run focused formatting, lint, and type checks; review changed code for concrete correctness and maintainability defects. Write JSONL findings plus final PASS/FAIL verdict to {report}, then run herdr-workflow verification-result --repo . --change {change} --role quality-verifier. No chat output." + persistent
    if role == "performance-verifier":
        context = f".herdr-workflow/{change}/reviews/round-{verification_round}-performance-verifier-context.md"
        report = f".herdr-workflow/{change}/reviews/round-{verification_round}-performance-verifier.findings.jsonl"
        return f"Silent performance verifier for {change} round {verification_round}. Read {context}. Review changed hot paths for measurable query, I/O, CPU, blocking, and memory regressions. Write JSONL findings plus final PASS/FAIL verdict to {report}, then run herdr-workflow verification-result --repo . --change {change} --role performance-verifier. No chat output." + persistent
    if role == "openspec-verifier":
        context = f".herdr-workflow/{change}/reviews/round-{verification_round}-openspec-verifier-context.md"
        report = f".herdr-workflow/{change}/reviews/round-{verification_round}-openspec-verifier.findings.jsonl"
        return f"Silent OpenSpec verifier for {change} round {verification_round}. Read {context}. Compare implementation against approved proposal, design, specs, and tasks for missing, incompatible, or out-of-scope behavior. Write JSONL findings plus final PASS/FAIL verdict to {report}, then run herdr-workflow verification-result --repo . --change {change} --role openspec-verifier. No chat output." + persistent
    if role == "test-verifier":
        context = f".herdr-workflow/{change}/reviews/round-{verification_round}-test-verifier-context.md"
        report = f".herdr-workflow/{change}/reviews/round-{verification_round}-test-verifier.findings.jsonl"
        return f"Silent full-suite test verifier for {change} round {verification_round}. Read {context}. Find and run repository's complete configured test suite without filters; PASS requires successful suite and regression coverage. Write JSONL findings plus final PASS/FAIL verdict to {report}, then run herdr-workflow verification-result --repo . --change {change} --role test-verifier. No chat output." + persistent
    if role == "archive":
        return f"Silent archive agent for OpenSpec change {change}. Read .herdr-workflow/{change}/reviews/archive-context.md only; do not read review history or telemetry. Follow its archive instructions, then run herdr-workflow archive --repo . --change {change}. No chat output." + restricted
    if role == "recovery":
        return f"Silent recovery agent for {change}. Read .herdr-workflow/{change}/reviews/recovery-context.json. Write .herdr-workflow/{change}/reviews/recovery-plan.json with matching recoveryId and exactly one allowlisted action: retry-verification, dispatch-triage, or record-verifier-result (include role). Do not execute plan, mutate state, commit, push, or archive. No chat output." + restricted
    raise ValueError(f"unknown role: {role}")


def pi_arguments(role, model, thinking, change, config):
    """Build Pi arguments for a Herdr-managed role agent."""
    skill = paths.SKILLS / f"herdr-openspec-{role}" / "SKILL.md"
    tools = ROLE_TOOLS[role]
    parts = ["--name", f"{change}-{role}", "--model", model, "--thinking", thinking]

    if role in UNRESTRICTED_ROLES:
        exclusions = resolve_exclusions(config, role)
        if exclusions:
            parts.append("--no-extensions")
            parts.extend(["--extension", str(paths.AGENT_DEF_DIR / "extensions" / "herdr-telemetry.ts")])
            parts.extend(["--extension", str(paths.AGENT_DEF_DIR / "extensions" / "herdr-workflow.ts")])
            for name, path in discover_extensions().items():
                if name not in exclusions and name not in HERDR_EXTENSIONS:
                    parts.extend(["--extension", path])
        else:
            parts.extend(["--extension", str(paths.AGENT_DEF_DIR / "extensions" / "herdr-telemetry.ts")])
            parts.extend(["--extension", str(paths.AGENT_DEF_DIR / "extensions" / "herdr-workflow.ts")])
        parts.extend(["--no-prompt-templates", "--skill", str(skill)])
    else:
        parts.extend([
            "--tools", tools, "--no-extensions",
            "--extension", str(paths.AGENT_DEF_DIR / "extensions" / "herdr-telemetry.ts"),
            "--extension", str(paths.AGENT_DEF_DIR / "extensions" / "herdr-workflow.ts"),
            "--no-prompt-templates", "--no-skills", "--skill", str(skill),
        ])

    if is_one_shot(role):
        parts.append("--no-session")
    if role == "archive":
        parts.append("--no-context-files")
    return parts

File: agentic-coding/src/workflow/prompts.ts
// Pure(ish) prompt/argument building. Only local filesystem reads for extension/plugin
// discovery — no subprocess.
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import * as paths from './paths.ts';
import * as naming from './naming.ts';
import { REPORT_CONTRACT } from './findings.ts';

export const UNRESTRICTED_ROLES = new Set(['planner', 'worker']);
export const ONE_SHOT_ROLES = new Set(['recovery', 'archive']);
// Herdr extensions loaded by explicit --extension flag, skip in discovery to avoid double-loading
export const HERDR_EXTENSIONS = new Set(['herdr-telemetry', 'herdr-workflow']);
export const PI_EXTENSION_DIRS = [path.join(paths.AGENT_DIR, 'extensions'), path.join(paths.AGENT_DEF_DIR, 'extensions'), path.join(os.homedir(), '.config', 'pi', 'extensions')];

export const ROLE_TOOLS: Record<string, string> = {
  planner: 'read,bash,edit,write',
  triage: 'read,bash,edit,write',
  recovery: 'read,bash,edit,write',
  worker: 'read,bash,edit,write',
  'security-verifier': 'read,bash',
  'agents-verifier': 'read,bash',
  'quality-verifier': 'read,bash',
  'performance-verifier': 'read,bash',
  'openspec-verifier': 'read,bash',
  'test-verifier': 'read,bash',
  'usability-verifier': 'read,bash',
  archive: 'read,bash',
};

export function isOneShot(role: string): boolean {
  return ONE_SHOT_ROLES.has(role);
}

export function roleEnv(role: string, change: string): string[] {
  return ['--env', `HERDR_ROLE=${role}`, '--env', `HERDR_CHANGE_ID=${change}`];
}

/** Discover all extension files from standard pi locations. */
export function discoverExtensions(): Record<string, string> {
  const extensions: Record<string, string> = {};
  for (const directory of PI_EXTENSION_DIRS) {
    if (!fs.existsSync(directory) || !fs.statSync(directory).isDirectory()) continue;
    for (const entry of fs.readdirSync(directory).sort()) {
      const ext = path.extname(entry);
      if (['.ts', '.js', '.mjs'].includes(ext)) {
        const name = path.basename(entry, ext);
        if (!(name in extensions)) extensions[name] = path.join(directory, entry);
      }
    }
  }
  return extensions;
}

/** Load plugin-assignments.json if it exists. */
export function loadPluginAssignments(): { plugins: Array<{ source: string; agentRoles: string[] }> } {
  const p = path.join(paths.AGENT_DIR, 'plugin-assignments.json');
  if (fs.existsSync(p)) {
    try {
      return JSON.parse(fs.readFileSync(p, 'utf8'));
    } catch {
      console.log(`warning: corrupt ${p}, starting fresh`);
    }
  }
  return { plugins: [] };
}

/** Save plugin-assignments.json. */
export function savePluginAssignments(assignments: unknown): void {
  const p = path.join(paths.AGENT_DIR, 'plugin-assignments.json');
  fs.writeFileSync(p, JSON.stringify(assignments, null, 2) + '\n');
}

/** Resolve excluded extension names for a given role. */
export function resolveExclusions(config: any, role: string): Set<string> {
  const excluded = new Set<string>();
  const plugins = config?.plugins ?? {};
  if (plugins.exclude_extensions) for (const name of plugins.exclude_extensions) excluded.add(name);
  const roleConfig = plugins.roles?.[role] ?? {};
  if (roleConfig.exclude_extensions) for (const name of roleConfig.exclude_extensions) excluded.add(name);
  return excluded;
}

export function rolePrompt(role: string, change: string, verificationRound?: number | null, workflowType?: string | null, task?: string | null): string {
  const request = `.herdr-workflow/${change}/request.md`;
  const restricted =
    ' Complete assigned role in this Pi process then stop — do not stay active waiting for next step. You will be notified when needed again. Do not invoke another agent executable or invoke another agent or pane. Use herdr-workflow for required workflow handoff exactly as specified.';
  const persistent =
    " Complete this round in this Pi process, then go idle and wait in this same process for the next round's prompt — do not exit, restart, or start unrelated work. Do not invoke another agent executable or invoke another agent or pane. Use herdr-workflow for required workflow handoff exactly as specified.";

  if (role === 'planner') {
    return `Planner for OpenSpec change ${change}. Read ${request}, explore repository context, and discuss unclear requirements with developer. When asked to propose, use openspec instructions proposal/design/tasks/specs --change ${change} as the artifact contract, then write proposal, design, tasks, and delta spec scenarios under openspec/changes/${change}/ and run openspec validate ${change} --strict. Do not call generic OpenSpec help or inspect archived changes solely to learn artifact structure. Submit with herdr-workflow phase proposed --repo . --change ${change}; fix PLAN_REJECTED feedback before finishing.`;
  }
  if (role === 'worker') {
    if (workflowType === 'no-openspec') {
      const description = task ? ` Implement this change: ${task}` : '';
      return `Worker for ${change}. Use chat for scope, progress, and blockers. Run focused tests only; quality and test verifiers own remaining validation gates. No task checklist to read — signal completion by running herdr-workflow verify --repo . --change ${change} once the change is applied. Output containing \`triage started:\` or \`verification already running:\` means handoff succeeded: stop immediately and never invoke verify again; do not poll status, call workflow help, inspect downstream workflow files, or wait with more tool calls. If a workflow phase/base/state blocker rejects it, report the exact blocker in chat and stop; never inspect workflow implementation source or read/change .herdr-workflow state (herdr.db).${description}`;
    }
    return `Worker for ${change}. Follow loaded skill and use chat for scope, progress, and blockers. Apply approved plan. Run focused tests only; quality and test verifiers own remaining validation gates. Mark each OpenSpec task [x] only after focused validation; verification rejects unfinished tasks. Output containing \`triage started:\` or \`verification already running:\` means handoff succeeded: stop immediately and never invoke verify again; do not poll status, call workflow help, inspect downstream workflow files, or wait with more tool calls. If a workflow phase/base/state blocker rejects it, report the exact blocker in chat and stop; never inspect workflow implementation source or read/change .herdr-workflow state (herdr.db).`;
  }
  if (role === 'triage') {
    const reviews = `.herdr-workflow/${change}/reviews`;
    return (
      `Silent triage for round ${verificationRound}. Read ${reviews}/round-${verificationRound}-triage-input.json only. Select minimum needed reviewers and assign relevant files or hunks. Do not select agents-verifier merely because AGENTS.md/CLAUDE.md applies; use it only when an instruction file changed or material tooling, environment, layout, or mandatory-command changes may require instruction updates. Write ${reviews}/round-${verificationRound}-triage.json, then run herdr-workflow dispatch-verifiers --repo . --change ${change}. No chat output.` +
      persistent
    );
  }
  const verifierFocus: Record<string, string> = {
    'security-verifier': 'Review changed trust boundaries for introduced injection, auth, secret, crypto, and input-validation defects.',
    'agents-verifier': 'Check changed code only against applicable AGENTS.md and CLAUDE.md instructions.',
    'quality-verifier': 'Run focused formatting, lint, and type checks; review changed code for concrete correctness and maintainability defects. Do not run any tests — test execution and coverage are the test verifier\'s domain.',
    'performance-verifier': 'Review changed hot paths for measurable query, I/O, CPU, blocking, and memory regressions.',
    'openspec-verifier': 'Compare implementation against approved proposal, design, specs, and tasks for missing, incompatible, or out-of-scope behavior.',
    'usability-verifier': 'Review changed frontend and asset files for introduced visual consistency, accessibility, responsive layout, design-system, component-state, and hardcoded-style defects.',
    'test-verifier': "Run the repository's complete configured test suite once without filters; do not rerun changed tests already covered. PASS requires success or only confirmed pre-existing unrelated failures. Reuse prior baseline evidence from context; one focused baseline reproduction is allowed only for a new apparently unrelated full-suite failure.",
  };
  const verifierLabel: Record<string, string> = {
    'security-verifier': 'security verifier',
    'agents-verifier': 'AGENTS instructions verifier',
    'quality-verifier': 'code quality verifier',
    'performance-verifier': 'performance verifier',
    'openspec-verifier': 'OpenSpec verifier',
    'usability-verifier': 'usability verifier',
    'test-verifier': 'full-suite test verifier',
  };
  if (role in verifierFocus) {
    const context = `.herdr-workflow/${change}/reviews/round-${verificationRound}-${role}-context.md`;
    const report = `.herdr-workflow/${change}/reviews/round-${verificationRound}-${role}.findings.jsonl`;
    const reviewOnly = !['quality-verifier', 'test-verifier'].includes(role)
      ? ' Review only; do not run tests, formatting, lint, type, or build commands. Treat generated context as authoritative scope: no full-repository git diff or unrelated changed-file inspection; read full assigned files or direct dependencies only when needed to understand a scoped hunk.'
      : '';
    return (
      `Silent ${verifierLabel[role]} for ${change} round ${verificationRound}. Read ${context}. ${verifierFocus[role]}${reviewOnly} ${REPORT_CONTRACT} Write the report to ${report}, then run herdr-workflow verification-result --repo . --change ${change} --role ${role}. No chat output. Only report actual defects and issues. Do not include findings that merely confirm code was implemented correctly — if nothing is wrong, write only the PASS verdict with no findings.` +
      persistent
    );
  }
  if (role === 'archive') {
    return (
      `Silent archive agent for OpenSpec change ${change}. Read .herdr-workflow/${change}/reviews/archive-context.md only; do not read review history or telemetry. Follow its archive instructions, then run herdr-workflow archive --repo . --change ${change}. No chat output.` +
      restricted
    );
  }
  if (role === 'recovery') {
    return (
      `Silent recovery agent for ${change}. Read .herdr-workflow/${change}/reviews/recovery-context.json. Write .herdr-workflow/${change}/reviews/recovery-plan.json with matching recoveryId and exactly one allowlisted action: retry-verification, dispatch-triage, or record-verifier-result (include role). Do not execute plan, mutate state, commit, push, or archive. No chat output.` +
      restricted
    );
  }
  throw new Error(`unknown role: ${role}`);
}

/** Build Pi arguments for a Herdr-managed role agent. */
export function piArguments(role: string, model: string, thinking: string, change: string, config: any): string[] {
  const skill = path.join(paths.SKILLS, `herdr-openspec-${role}`, 'SKILL.md');
  const tools = ROLE_TOOLS[role];
  const parts = ['--name', naming.agentName(change, role), '--model', model, '--thinking', thinking];
  const telemetryExt = path.join(paths.AGENT_DEF_DIR, 'extensions', 'herdr-telemetry.ts');
  const workflowExt = path.join(paths.AGENT_DEF_DIR, 'extensions', 'herdr-workflow.ts');

  if (UNRESTRICTED_ROLES.has(role)) {
    const exclusions = resolveExclusions(config, role);
    if (exclusions.size) {
      parts.push('--no-extensions');
      parts.push('--extension', telemetryExt);
      parts.push('--extension', workflowExt);
      for (const [name, extPath] of Object.entries(discoverExtensions())) {
        if (!exclusions.has(name) && !HERDR_EXTENSIONS.has(name)) parts.push('--extension', extPath);
      }
    } else {
      parts.push('--extension', telemetryExt);
      parts.push('--extension', workflowExt);
    }
    parts.push('--no-prompt-templates', '--skill', skill);
  } else {
    parts.push('--tools', tools, '--no-extensions', '--extension', telemetryExt, '--extension', workflowExt, '--no-prompt-templates', '--no-skills', '--skill', skill);
  }

  if (isOneShot(role)) parts.push('--no-session');
  if (role === 'archive') parts.push('--no-context-files');
  return parts;
}
